transport-selector/src/main/java/io/netty/transport/selector/TransportSelector.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.transport.selector;
+
+import io.netty.channel.Channel;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.epoll.Epoll;
+import io.netty.channel.epoll.EpollDatagramChannel;
+import io.netty.channel.epoll.EpollDomainSocketChannel;
+import io.netty.channel.epoll.EpollEventLoopGroup;
+import io.netty.channel.epoll.EpollServerDomainSocketChannel;
+import io.netty.channel.epoll.EpollServerSocketChannel;
+import io.netty.channel.epoll.EpollSocketChannel;
+import io.netty.channel.kqueue.KQueue;
+import io.netty.channel.kqueue.KQueueDatagramChannel;
+import io.netty.channel.kqueue.KQueueDomainSocketChannel;
+import io.netty.channel.kqueue.KQueueEventLoopGroup;
+import io.netty.channel.kqueue.KQueueServerDomainSocketChannel;
+import io.netty.channel.kqueue.KQueueServerSocketChannel;
+import io.netty.channel.kqueue.KQueueSocketChannel;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.DatagramChannel;
+import io.netty.channel.socket.ServerSocketChannel;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioDatagramChannel;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+import io.netty.channel.socket.nio.NioSocketChannel;
+import io.netty.channel.unix.DomainSocketChannel;
+import io.netty.channel.unix.ServerDomainSocketChannel;
+
+import java.util.concurrent.ThreadFactory;
+
+/**
+ * Helper class which will hand you the <strong>optimal</strong> implementation of a transport depending on the
+ * used platform and the native libraries on the classpath.
+ *
+ * It is important to note that if you use any of the methods to select the right {@link Channel} you should also use
```

Comment:
In order to make this explicit; you can consider returning a "tuple" from the methods instead of channel and eventloop group individually, eg:

```java

    public static final class Tuple<T extends Channel, EG extends EventLoopGroup> {
        private final Class<? extends T> channel;
        private final EG eventloopGroup;

        private Tuple(final Class<? extends T> channel, final EG eventloopGroup) {
            this.channel = channel;
            this.eventloopGroup = eventloopGroup;
        }

        public Class<? extends T> getChannel() {
            return channel;
        }

        public EG getEventloopGroup() {
            return eventloopGroup;
        }

        public void configureBootstrap(AbstractBootstrap<?, T> bootstrap) {
            bootstrap.channel(channel);
            bootstrap.group(eventloopGroup);
        }
    }

    public static Tuple<SocketChannel, EventLoopGroup> socket() {
        Class<? extends SocketChannel> c = socketChannel();
        EventLoopGroup eg = newEventLoopGroup();
        return new Tuple<>(c, eg);
    }

    public static Tuple<ServerSocketChannel, EventLoopGroup> serverSocket() {
        Class<? extends ServerSocketChannel> c = serverSocketChannel();
        EventLoopGroup eg = newEventLoopGroup();
        return new Tuple<>(c, eg);
    }
```

This also provides an opportunity to help setup the bootstrap as demonstrated above in `Tuple#configureBootstrap()`

A negative is that the number of static methods will have a combinatorial factor based on arguments to create an eventloop group but they would not be out of control 😄

Reply:
this is a bit of a nit (in response to the above):
Why not just create a small class to hold all of them.

```java
TransportSelections selections = TransportSelector.selectOptionalForNative();
```

Reply:
Something like `TransportSelections`, `TransportStrategy`, `TransportSettings`, `TransportConfiguration`, etc. might make sense. It might also lend itself to being more DI-friendly.

However, rather than `Tuple#configureBootstrap(Bootstrap)`, I think the opposite direction would be more user-friendly, something like `Bootstrap#configure(Tuple)`.

Reply:
I think tuple is a bit of an anti-pattern on the GC side of things and it's not commonly used by netty, unless it's some kind of `static final` variables.